GpsDataAnalyzer/fileparser/fileparser.py

```diff
@@ -46,14 +50,115 @@ def parse_file(self, filename) -> GpsDataSet:
         file_type = self.get_file_type(filename)
         
         if file_type == 'xml':
+            logger.info('Parse xml file ' + filename)
             return self.parse_xml(filename)
         elif file_type == 'csv':
+            logger.info('Parse csv file ' + filename)
             return self.parse_csv(filename)
         else:
-            print("Invalid file type!")
+            logger.debug('Invalid file type.')
             return None
 
 
+    def parse_time(self, timestr) -> datetime:
+        """
+        Parse the xml timestamp and return a new datetime object
+
+        Args:
+          timestr: string of the timestamp extracted from xml file
+
+        Returns:
+          A datetime object with timezone
+        """
+        #Replace timezone Z with '' since python datetime parser could not handler %Z
+        timestr = timestr.replace('Z', '', 1)
+        #Parse time in the format "%Y-%m-%dT%H:%M:%S.%f"
+        time = datetime.strptime(timestr, "%Y-%m-%dT%H:%M:%S.%f")
+        #Set timezone back
+        tz_time = time.replace(tzinfo=timezone.utc)
+
+        return tz_time
+
+
+    def calculate_distance(self, location1, location2) -> float:
+        """
+        Calculate 3D distance (including altitude) between two location points with ellipsoidal earth model
+
+        Args:
+          location1: tuple of (latitude, longitude) as floats in Decimal Degrees
+          location2: tuple of (latitude, longitude) as floats in Decimal Degrees
+
+        Returns:
+           A float in meters of the distance between two location points
+        """
+        return geodesic(location1, location2).meters
+
+
+    def parse_xml_metadata(self, root, prefix) -> GpsMetaData:
+        """
+        Helper function to parse metadata information in xml file
+        """
+        #Get metadata information
+        metadata = root.find(prefix + 'metadata')
+        # Parse start timestamp
+        starttimestr = metadata.find(prefix + 'time').text
```

Comment:
any clue what's the complexity of the metadata.find function? Is it just a linear search or something better?

The reason I'm asking is for scalability reasons - how well does this program work as the filesize of the GPS XML scales?

Reply:
That's a good question. Overall speaking, ElementTree needs less memory for XML trees than DOM and thus is faster. The `ElementTree parse() function` returns a tree structure of lists, and attributes are represented as dictionaries. So `root.find()` should be in constant time O(h), h refers to the height of tree, `metadata.find()` and `trkpt.find()` should be  O(n), n refers to the length of metadata list, and the `trkpt.get()` should be O(1) since the attributes are presented as dictionaries. 

From my understanding, the metadata list length won't change much (it only contains those necessary identifiers) Which did influence the performance of the parsing is the `trkpt parsing` while it scales. Since we need to parse each data point, a linear iteration over the `trkseg` is the best I could think about. If you have better approach, please give me some hint. :)

Reply:
oh sorry I didn't realize this was metadata only, so I am overthinking this a bit. 

It's a small optimization I think, but may not be worthwhile to code it. 

This is what I was thinking, not sure if it actually works though since I didn't implement it --

metadata.find() is called 5 times, with a worse case scenario of 5*O(n). We could potentially reduce the constant down to 1*O(n), by using an iterator:
for data in metadata.iter():
  if (data.key.contains('device'))
    device = data.text
  elif (data.key.contains('id'))
    identifier = data.text
...etc

probably makes sense if n is LARGE or we have to iterate through the list many many times.
So just an FYI for here and perhaps it might be useful for the future.

Reply:
That make sense, let me try that out. Thanks!